docs/proposals/acting-as.md

```diff
@@ -0,0 +1,219 @@
+# Proposal: Solving the Confused Deputy Problem with Acting-As
+
+## Proposal
+
+**Ongoing Problem**: we have had confused deputy bugs in RC-manager, job-manager, and HPA-manager.
+It is an insidious problem in general.  We need to attenuate the authority of controller-manager.
+
+**Core Kubernetes Principle**: controllers from the core project, like replication controller, are
+not special -- they are on a level playing field with user-implemented components.  (Only apiservers
+are special, and they should do as little work as practical.)
+
+**Observation**:  Instead of POSTing an RC and letting the RC-manager watch it for me, I could write a
+go program that watches my pods, and makes new ones for me, emulating an RC.  Currently the auth
+flow for those two cases is totally different.  But it should not be very different, (see above
+Principle).  
+
+This diagram shows how it works today when a user creates an RC, and then the RC-manager
+creates a pod.
+
+<img src="https://www.lucidchart.com/publicSegments/view/14d7b49f-469c-4bc4-8bf7-f13581ee0220/image.png"
+alt="UML Sequence Diagram for Current Replication Controller">
+<!-- Original document: https://www.lucidchart.com/documents/edit/34454705-cff6-4b42-a95f-24d69a36184c
+     contact etune@google.com to request access -->
+
+This diagram shows how it woud work today if a user created a program that was itself a Kubernetes
+client, and acted as an RC, but for a fixed template, number of replicas, etc.
+
+<img src="https://www.lucidchart.com/publicSegments/view/a719bd66-191f-478f-8c42-bc033d19ef25/image.png"
+alt="UML Sequence Diagram for User-written Replication Controller">
+<!-- Original document: https://www.lucidchart.com/documents/edit/92c9f481-a104-4283-8c8d-6bec71a1d9d7
+     contact etune@google.com to request access -->
+
+This diagram shows a proposal for how I think RC-manager should work.  It uses a new header (or
+maybe query parameter?) which tells the apiserver that the RC-manager wants to Act-As
+a different user. It is much closer to the "user-written replication manager" case.
+
+<img src="https://www.lucidchart.com/publicSegments/view/c59b1964-2345-4910-8580-d4c636f28fa3/image.png"
+alt="UML Sequence Diagram for Proposed RC-manager">
+<!-- Original document: https://www.lucidchart.com/documents/edit/7c8cdcde-9803-4211-8cf1-a05c6da3fbc6
+     contact etune@google.com to request access -->
+
+Things to note in the new flow:
+
+- controller-manager is using authorized port instead of the localhost port to talk to 
+   kube-apiserver
+- controller-manager sets a header to say it wants to lower its root-like powers in order
+  to act as less powerful user (which solves the confused deputy).
```

Comment:
It could help with the confused deputy, but it depends when it gets confused... if the problem is that it gets tricked into doing something cross-namespace, it could correctly lower its permissions to scope to the namespace it was tricked into, but it would be too late...
Do you envision changing controllers to operate on a single namespace at a time? That seems more likely to help, but also have a much larger performance impact (N list operations instead of a single one across all namespaces, etc)